autogen/main/cluster.tf.tmpl

```diff
@@ -266,7 +266,7 @@ resource "google_container_cluster" "primary" {
   enable_autopilot = true
 {% endif %}
   dynamic "master_authorized_networks_config" {
-    for_each = {% if private_cluster %}var.enable_private_endpoint || {% endif %}var.gcp_public_cidrs_access_enabled != null || length(var.master_authorized_networks) > 0 ? [true] : []
+    for_each = {% if private_cluster %}(var.enable_private_endpoint && !var.deploy_using_private_endpoint) || {% endif %}var.gcp_public_cidrs_access_enabled != null || length(var.master_authorized_networks) > 0 ? [true] : []
```

Comment:
Based on https://github.com/hashicorp/terraform-provider-google/blob/main/google/services/container/resource_container_cluster.go#L5035, I'm thinking we should probably gate on `master_authorized_networks` being not `null`.  That should more properly differentiate between `null` (no `master_authorized_networks_config` block, hence disable) and an empty list `[]` (with `master_authorized_networks_config` block enabled, but no `cidr_blocks`):

```suggestion
    for_each = var.master_authorized_networks != null ? [true] : []
```

Reply:
I updated the PR. This works for me; however, I have to explicitly unset `master_authorized_networks` because the default is `[]`, but that's okay.

```
enable_private_endpoint       = true
deploy_using_private_endpoint = true
master_authorized_networks    = null
```

Reply:
Yeah - Ideally we would change the `master_authorized_networks` default to `null` and update the description, but that would be a breaking change.  That can still be done in the future, but we should be able to ship this as just a fix for now.

Reply:
Actually @TheKangaroo this will also result in a breaking change of behavior for clusters which didn't have one of the three previously set `var.enable_private_endpoint || var.gcp_public_cidrs_access_enabled != null || length(var.master_authorized_networks) > 0`, as `master_authorized_networks_config {}` would newly be created for them.

I think we have a couple options:
- Update the `master_authorized_networks` default to `null`. (breaking change, but following the TPG behavior) <- Probably the "best" option?
- Add a new `master_authorized_networks_enabled` which defaults to `true` and would instead be used to gate this block. (more closely follow the API behavior) <- Prefer not to add new variables if possible.
- Something else?

Reply:
I don't think I need to add an explanation in the variable description on why the default is null, but we should include an explanation in the release notes/update docs. However, I do not know where the release-bot will pull this information from. Is it from the PR description or from the commit message?

Reply:
@apeabody I added an upgrade note to the PR, basing it on the default change notice from the last major release.